internal/netsync/manager.go

```diff
@@ -1335,167 +1303,126 @@ func (m *SyncManager) handleInvMsg(imsg *invMsg) {
-				gdmsg.AddInvVect(iv)
-				numRequested++
+				requestQueue = append(requestQueue, iv)
 			}
 
 		case wire.InvTypeTx:
-			// Request the transaction if there is not already a
-			// pending request.
+			// Add the tx to the cache of known inventory for the peer.  This
+			// helps avoid sending transactions to the peer that it is already
+			// known to have.
+			peer.AddKnownInventory(iv)
+
+			// Ignore transaction announcements before the chain is current or
+			// are otherwise not needed, such as when they were recently
+			// rejected or are already known.
+			//
+			// Transaction announcements are based on the state of the fully
+			// synced ledger, so they are likely to be invalid before the chain
+			// is current.
+			if !isCurrent || !m.needTx(&iv.Hash) {
+				continue
+			}
+
+			// Request the transaction if there is not one already pending.
 			if _, exists := m.requestedTxns[iv.Hash]; !exists {
 				limitAdd(m.requestedTxns, iv.Hash, maxRequestedTxns)
 				limitAdd(peer.requestedTxns, iv.Hash, maxRequestedTxns)
-				gdmsg.AddInvVect(iv)
-				numRequested++
+				requestQueue = append(requestQueue, iv)
 			}
 		}
+	}
 
+	if lastBlock != nil {
+		// The block is an orphan block that we already have.  When the
+		// existing orphan was processed, it requested the missing parent
+		// blocks.  When this scenario happens, it means there were more
+		// blocks missing than are allowed into a single inventory message.
+		// As a result, once this peer requested the final advertised block,
+		// the remote peer noticed and is now resending the orphan block as
+		// an available block to signal there are more missing blocks that
+		// need to be requested.
+		if m.isKnownOrphan(&lastBlock.Hash) {
+			// Request blocks starting at the latest known up to the root of the
+			// orphan that just came in.
+			orphanRoot := m.orphanRoot(&lastBlock.Hash)
+			blkLocator := m.cfg.Chain.LatestBlockLocator()
+			locator := chainBlockLocatorToHashes(blkLocator)
+			err := peer.PushGetBlocksMsg(locator, orphanRoot)
+			if err != nil {
+				log.Errorf("Failed to push getblocksmsg for orphan chain: %v",
+					err)
+			}
+		}
+
+		// Update the last announced block to the final one in the announced
+		// inventory above (if any).  In the case the header for that block is
+		// already known, use that information to update the height for the peer
+		// too.
+		peer.UpdateLastAnnouncedBlock(&lastBlock.Hash)
+		if isCurrent {
+			header, err := m.cfg.Chain.HeaderByHash(&lastBlock.Hash)
+			if err == nil {
+				peer.UpdateLastBlockHeight(int64(header.Height))
+			}
+		}
+
+		// Request blocks after the last one advertised up to the final one the
+		// remote peer knows about (zero stop hash).
+		blkLocator := m.cfg.Chain.BlockLocatorFromHash(&lastBlock.Hash)
+		locator := chainBlockLocatorToHashes(blkLocator)
+		err := peer.PushGetBlocksMsg(locator, &zeroHash)
+		if err != nil {
+			log.Errorf("PEER: Failed to push getblocksmsg: %v", err)
```

Comment:
Super minor, but here the error log is prefixed with `PEER:` where it is not for other cases, and it would probably be best to keep it consistent.